query/src/main/java/org/infinispan/query/affinity/AffinityIndexManager.java

```diff
@@ -0,0 +1,286 @@
+            if (throwable == null) {
+               logDebug("Operation completed");
+            } else {
+               log.errorf(throwable, "Error reapplying operation");
+            }
+         });
+      }
+
+   }
+
+   private void clearIfNeeded(Address lockHolder) {
+      if(!rpcManager.getMembers().contains(lockHolder)) {
+         final Directory directory = getDirectoryProvider().getDirectory();
+         log.warn("Forcing clear of index lock");
+         ((DirectoryExtensions) directory).forceUnlock(IndexWriter.WRITE_LOCK_NAME);
+      }
+   }
+
+   @Override
+   public void initialize(String indexName, Properties properties, Similarity similarity, WorkerBuildContext buildContext) {
+      super.initialize(indexName, properties, similarity, buildContext);
+      ServiceManager serviceManager = buildContext.getServiceManager();
+      ComponentRegistryService componentRegistryService = serviceManager.requestService(ComponentRegistryService.class);
+      ComponentRegistry componentRegistry = componentRegistryService.getComponentRegistry();
+      distributionManager = componentRegistry.getComponent(DistributionManager.class);
+      rpcManager = componentRegistry.getComponent(RpcManager.class);
+      cache = componentRegistry.getComponent(Cache.class);
+      keyTransformationHandler = componentRegistry.getComponent(QueryInterceptor.class).getKeyTransformationHandler();
+      SearchIntegrator component = componentRegistry.getComponent(SearchIntegrator.class);
+      AffinityErrorHandler errorHandler = (AffinityErrorHandler) component.getErrorHandler();
+      cache.addListener(new TopologyChangeListener());
+      errorHandler.initialize(this);
+   }
+
+   private void handleOwnershipLost(Address newOwner) {
+      writeLock.lock();
+      try {
+         logDebug("Ownership lost, closing index manager");
+         nextOwner = newOwner;
+         flushAndReleaseResources();
+         hasOwnership = false;
+      } finally {
+         writeLock.unlock();
+      }
+   }
+
+   private void handleOwnershipAcquired() {
+      writeLock.lock();
+      try {
+         logDebug("Ownership acquired");
+         nextOwner = null;
+         hasOwnership = true;
+      } finally {
+         writeLock.unlock();
+      }
+   }
+
+   @Override
+   public void flushAndReleaseResources() {
+      Address lockHolder = getLockHolder(getIndexName());
+      if (lockHolder != null) {
+         InfinispanDirectoryProvider directoryProvider = (InfinispanDirectoryProvider) getDirectoryProvider();
+         int activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
+         while (activeDeleteTasks > 0) {
+            try {
+               Thread.sleep(POLL_WAIT);
+               logDebug("Waiting for pending delete tasks, remaining: %s", activeDeleteTasks);
+               activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
+            } catch (InterruptedException e) {
+               Thread.currentThread().interrupt();
```

Comment:
This will get in an infinite loop if it ever happens :-1: 
If we want to ignore the interrupt we need to set a boolean and then reset the interrupt flag after we are outside of the while loop.